keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+...
+spec:
+  paramSource:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    ...
+  validations:
+    expression: "object.spec.replicas <= params.maxReplicas"
+    name: max-replicas
+  defaultValidations:
+  - expression: "object.spec.replicas <= 5"
+    enforcement: [Deny]
+```
+
+Reporting/debugging/analysis implications:
+
+- Violations for non-singleton policies will always be reported using a {policy
+  definition, policy binding} identifier pair. To be consistent with this,
+  singleton policies can be use a {policy definition, policy definition}
+  identifier pair. This is a bit verbose but keeps reporting consistent and
+  makes tracing a violation back to resources that produced it straight-forward.
+
+### Phase 2
+
+All these capabilities are required before Beta, but will not be implemented
+in the first alpha release of this enhancement due to the size and complexity
+of this enhancement.
+
+#### Namespace scoped policy binding
+
+For phase 1, policy bindings were only allowed to be cluster scoped. We can
+support namespace scoped policy bindings as follows:
+
+- Add a `NamespacePolicyBinding` resource.
+- If the parameter resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. As an example, ResourceQuota works this way.
+
+Details to consider:
+
+- Should a policy support both cluster scoped and namespace scoped binding? If
+  so how? It would need two different parameter CRDs (since a CRD must either be
+  cluster scoped or namespace scoped, not both).
+
+#### CEL Expression Composition
+
+##### Variables
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+cel-policy-template "terms" are very convenient. But we don't have consensus on
+if they are sufficiently essential. Need to either agree on something or drop
+this.
+<<[/UNRESOLVED]>>
+
+Each CEL "program" is a single expression. There is no support for vaiable
+assignment. This can result in redundant code to traverse maps/arrays or
+dereference particular fields.
+
+We can support this in much the same way as cel-policy-template if we offer an
+ordered list or variables. These can be lazily evaluated while the validation
+expressions are evaluated (cel-policy-template does this). The results can also
+be memoized to avoid repeated evaluations if they are shared across validations.
+
+```yaml
+  variables:
+    - name: metadataList
+      expression: "spec.list.map(x, x.metadata)"
+    - name: itemMetadataNames
+      expression: "metadataList.map(m, m.name)"
+  validations:
+    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "itemMetadataNames.exists(name, name == 'required')"
+```
+
+##### Scopes
+
+<<[UNRESOLVED jpbetz, ?? ]>> 
+This is not yet agreed on, and some reviewers have found it more
```

Comment:
IMO I would include "bind variables" and omit "scopes", since they share functionality but scopes is less flexible than bind variables and adds a second path traversal language.

Having some kind of forEach for binding variables might be necessary to simplify expressions that traverse nested collections (e.g. all ports in all containers).

Reply:
I at least agree you should pick one way to do this, not two or three.

Reply:
I'll focus on "variables" and drop scopes. This is also phase 2 so we have an additional cycle to refine the proposal.